compiler/rustc_infer/src/traits/mod.rs

```diff
@@ -105,6 +109,65 @@ pub enum FulfillmentErrorCode<'tcx> {
     CodeAmbiguity,
 }
 
+pub struct ObligationsDedup<'a, 'tcx, T> {
+    obligations: &'a mut Vec<Obligation<'tcx, T>>,
+}
+
+impl<'a, 'tcx, T: 'tcx> ObligationsDedup<'a, 'tcx, T>
+where
+    T: Clone + Hash + Eq,
+{
+    pub fn from_vec(vec: &'a mut Vec<Obligation<'tcx, T>>) -> Self {
+        ObligationsDedup { obligations: vec }
+    }
+
+    pub fn extend<'b>(&mut self, iter: impl ExactSizeIterator<Item = Cow<'b, Obligation<'tcx, T>>>)
+    where
+        'tcx: 'b,
+    {
+        // obligation tracing has shown that initial batches added to an empty vec do not
+        // contain any duplicates, so there's no need to attempt deduplication
+        if self.obligations.is_empty() {
+            *self.obligations = iter.into_iter().map(Cow::into_owned).collect();
+            return;
+        }
+
+        let initial_size = self.obligations.len();
+        let current_capacity = self.obligations.capacity();
+        let iter = iter.into_iter();
+        let expected_new = iter.len();
+        let combined_size = initial_size + expected_new;
+
+        if combined_size <= 16 || combined_size <= current_capacity {
+            // small case/not crossing a power of two. don't bother with dedup
+            self.obligations.extend(iter.map(Cow::into_owned));
+        } else {
+            // crossing power of two threshold. this would incur a vec growth anyway if we didn't do
+            // anything. piggyback a dedup on that
+            let mut seen = FxHashMap::default();
+            seen.reserve(initial_size);
+
+            let mut is_duplicate = move |obligation: &Obligation<'tcx, _>| -> bool {
+                return match seen.raw_entry_mut().from_key(obligation) {
+                    RawEntryMut::Occupied(..) => true,
+                    RawEntryMut::Vacant(vacant) => {
+                        vacant.insert(obligation.clone(), ());
+                        false
+                    }
+                };
+            };
+
+            self.obligations.retain(|obligation| !is_duplicate(obligation));
```

Comment:
I know that we don't dedup earlier because of cache effects (a perf optimization), but doesn't that mean that you *always* incur the O(n) cost of deduplicating the obligations caught here? Can we check what the perf implications would be of having state in the `struct` for "`self.obligations` is already deduped"?

Reply:
> but doesn't that mean that you always incur the O(n) cost of deduplicating the obligations caught here?

Well, not always the vec is grown, no, that would result in O(n²) time. It's amortized O(n) because we only do it when the vec would grow.

> Can we check what the perf implications would be of having state in the struct for "self.obligations is already deduped"?

Currently the dedup struct is only constructed temporarily in method scope, it lives much shorter than the vec, so it can't remember whether the vec has already been deduped. It would require larger changes to plumb the dedup state through all method calls that pass the `&mut Vec<Obligations>` around. That's what I meant with the following in the PR description

> or we need to start plumbing the deduplication or a sorted set all the way through the APIs that currently use Vec.